drv/lpc55-swd/src/main.rs

```diff
@@ -392,91 +525,43 @@ impl idl::InOrderSpCtrlImpl for ServerImpl {
         &mut self,
         _: &RecvMessage,
     ) -> Result<(), RequestError<SpCtrlError>> {
-        if !self.init {
-            self.pin_setup();
-        }
-
-        if self.swd_dongle_detected() {
-            ringbuf_entry!(Trace::DongleDetected);
-            return Err(SpCtrlError::DongleDetected.into());
-        }
-
-        match self.swd_setup() {
-            Ok(_) => {
-                self.init = true;
-                Ok(())
-            }
-            Err(_) => Err(SpCtrlError::Fault.into()),
-        }
+        self.do_setup_swd()?;
+        Ok(())
     }
 
     fn halt(
         &mut self,
         _: &RecvMessage,
     ) -> Result<(), RequestError<SpCtrlError>> {
-        match self.write_single_target_addr(DHCSR, DHCSR_HALT_MAGIC) {
-            Ok(_) => loop {
-                match self.read_single_target_addr(DHCSR) {
-                    Ok(dhcsr) => {
-                        ringbuf_entry!(Trace::Dhcsr(dhcsr));
-
-                        if dhcsr & DHCSR_S_HALT != 0 {
-                            return Ok(());
-                        }
-                    }
-                    Err(_) => {
-                        return Err(SpCtrlError::Fault.into());
-                    }
-                }
-            },
-            Err(_) => Err(SpCtrlError::Fault.into()),
-        }
+        self.do_halt()?;
+        Ok(())
     }
 
     fn resume(
         &mut self,
         _: &RecvMessage,
     ) -> Result<(), RequestError<SpCtrlError>> {
-        match self.write_single_target_addr(DHCSR, DHCSR_RESUME_MAGIC) {
-            Ok(_) => Ok(()),
-            Err(_) => Err(SpCtrlError::Fault.into()),
-        }
+        self.do_resume()?;
+        Ok(())
     }
 
     fn read_core_register(
         &mut self,
         _: &RecvMessage,
         register: u16,
     ) -> Result<u32, RequestError<SpCtrlError>> {
-        // C1.6 Debug system registers
-        let r = match register {
-            // R0-R12
-            0b0000000..=0b0001100
-
-            // LR - PSP
-            | 0b0001101..=0b0010010
-
-            // CONTROL/FAULTMASK/BASEPRI/PRIMASK
-            | 0b0010100
-
-            // FPCSR
-            | 0b0100001
-
-            // S0-S31
-            | 0b1000000..=0b1011111 => Ok::<u16, SpCtrlError>(register),
-            _ => Err(SpCtrlError::InvalidCoreRegister)
-        }?;
-
-        if self.write_single_target_addr(DCRSR, r as u32).is_err() {
-            return Err(SpCtrlError::Fault.into());
-        }
-
+        Reg::from_u16(register)
+            .ok_or(SpCtrlError::InvalidCoreRegister)
+            .and_then(|r| {
+                self.write_single_target_addr(DCRSR, r as u32)
+                    .map_err(|_| SpCtrlError::Fault)
+            })?;
```

Comment:
vibes: I would break this into two operations, instead of chaining with `and_then`
```suggestion
        let r = Reg::from_u16(register)
            .ok_or(SpCtrlError::InvalidCoreRegister)?;
        self.write_single_target_addr(DCRSR, r as u32)
            .map_err(|_| SpCtrlError::Fault)?;
```

Reply:
done